src/wasm/wasm-type.cpp

```diff
@@ -1180,6 +1213,221 @@ bool SubTyper::isSubType(const Rtt& a, const Rtt& b) {
+  Type tempLUB;
+  if (!lub(a, b, tempLUB)) {
+    return Type::none;
+  }
+  // `tempLUB` is a temporary type owned by `builder`. Since TypeBuilder::build
+  // returns HeapTypes rather than Types, create a new HeapType definition meant
+  // only to get `tempLUB` canonicalized in a known location. The use of an
+  // Array is arbitrary; it might as well have been a Struct.
+  builder.grow(1);
+  builder[builder.size() - 1] = Array(Field(tempLUB, Mutable));
+  return builder.build().back().getArray().element.type;
+}
+
+bool TypeBounder::lub(Type a, Type b, Type& out) {
+  if (a == b) {
+    out = a;
+    return true;
+  }
+  if (a == Type::unreachable) {
+    out = b;
+    return true;
+  }
+  if (b == Type::unreachable) {
+    out = a;
+    return true;
+  }
+  if (a.isTuple() && b.isTuple()) {
+    Tuple tuple;
+    if (!lub(a.getTuple(), b.getTuple(), tuple)) {
+      return false;
+    }
+    out = builder.getTempTupleType(tuple);
+    return true;
+  } else if (a.isRef() && b.isRef()) {
+    auto nullability =
+      (a.isNullable() || b.isNullable()) ? Nullable : NonNullable;
+    HeapType heapType = lub(a.getHeapType(), b.getHeapType());
+    out = builder.getTempRefType(heapType, nullability);
+    return true;
+  } else if (a.isRtt() && b.isRtt()) {
+    Rtt rtt(HeapType::any);
+    if (!lub(a.getRtt(), b.getRtt(), rtt)) {
+      return false;
+    }
+    out = builder.getTempRttType(rtt);
+    return true;
+  }
+  return false;
+}
+
+HeapType TypeBounder::lub(HeapType a, HeapType b) {
+  if (a == b) {
+    return a;
+  }
+  // Canonicalize to have the basic HeapType on the left.
+  if (b.isBasic()) {
+    std::swap(a, b);
+  }
+  if (a.isBasic()) {
+    switch (a.getBasic()) {
+      case HeapType::func:
+        if (b.isFunction()) {
+          return HeapType::func;
+        } else {
+          return HeapType::any;
+        }
+      case HeapType::ext:
+        return HeapType::any;
+      case HeapType::any:
+        return HeapType::any;
+      case HeapType::eq:
+        if (b == HeapType::i31 || b.isData()) {
+          return HeapType::eq;
+        } else {
+          return HeapType::any;
+        }
+      case HeapType::i31:
+        if (b == HeapType::data || b.isData()) {
+          return HeapType::eq;
+        } else {
+          return HeapType::any;
+        }
+      case HeapType::data:
+        if (b.isData()) {
+          return HeapType::data;
+        } else if (b == HeapType::i31) {
+          return HeapType::eq;
+        } else {
+          return HeapType::any;
+        }
+    }
+  }
+  // Allocate a new slot to construct the LUB of this pair if we have not
+  // already seen it before. Canonicalize the pair to have the element with the
+  // smaller ID first since order does not matter.
+  auto pair =
+    a.getID() < b.getID() ? std::make_pair(a, b) : std::make_pair(b, a);
+  size_t index = builder.size();
+  auto result = indices.insert({pair, index});
+  if (result.second) {
+    builder.grow(1);
```

Comment:
I'm confused. How can this not be a new pair every time? `index` is the size of the builder, which increases monotonically each time we arrive here.

Reply:
Oh, right, I think I was confused by `insert` receiving a pair while the key is itself a pair...